src/RTMClient.ts

```diff
@@ -231,10 +233,11 @@ export class RTMClient extends EventEmitter {
           .onSuccess().transitionTo('connecting')
         .onExit(() => this.teardownWebsocket())
       .global()
-        .onStateEnter((state) => {
+        .onStateEnter((state, context) => {
           this.logger.debug(`transitioning to state: ${state}`);
           // Emits events: `disconnected`, `connecting`, `connected`, 'disconnecting', 'reconnecting'
-          this.emit(state);
+          // event payload is anything related to the event, i.e. an error on disconnect
+          this.emit(state, context.eventPayload);
```

Comment:
it seems like this could have effects greater than passing through the error on disconnect. i'd like to be a little more explicit about what this argument will contain for each kind of transition.

Reply:
Just looking through the code, the only other case I could see where something else passes through would be the [event payloads for web socket events](https://github.com/slackapi/node-slack-sdk/blob/b207d885cae73aef97301de8b1b65f7b1195ffb7/src/RTMClient.ts#L554-L555), specifically the [non-existent `open` event's payload](https://github.com/websockets/ws/blob/e4d032c383dd2931de7dfbe81a0a8185cdffbf52/doc/ws.md#event-open) and the web socket closing code and reason from the [`close` event](https://github.com/websockets/ws/blob/e4d032c383dd2931de7dfbe81a0a8185cdffbf52/doc/ws.md#event-close).

I agree, though, that this could end up leaking something else through. Is there a better course of action? Like guarding that only `eventPayload instanceof Error` passes?

Reply:
it seems like the intention is to _only_ emit the `context.eventPayload` when transitioning to `"disconected". if that's true, i think it would be clearer to do something like:

```javascript
if (state === 'disconnected') {
  // when arriving at this state due to failure, the eventPayload will contain an error.
  // otherwise, like after a force disconnect, this value will be undefined
  const errorOrUndefined = context.eventPayload;
  this.emit(state, errorOrUndefined);
} else {
  this.emit(state);
}
```

its more explicit, and gives us a chance to document the intention.